hail/python/hail/ggplot/scale.py

```diff
@@ -212,13 +212,28 @@ def get_values(self, categories):
 
 
 class ScaleColorContinuous(ScaleContinuous):
+    def is_discrete_output(self):
+        return True
```

Comment:
I'm not sure what distinction you're making with the `is_discrete`->`is_discrete_output` change. I think it's pretty fundamental in the ggplot model that we have discrete and continuous scales, and they behave differently. Making `ScaleColorContinuous` a discrete scale seems confusing at best. We do have `ScaleColorHue` as a discrete color scale. Can you explain the motivation for this change?

Reply:
Ahh, hmm. So this was a great comment. It sent me down a rabbit hole and I realized I misunderstand (and also our users misunderstand) what `scale_color_continuous` does.

We have two color scales:

- `scale_color_hue` (aliased by `scale_color_discrete`) which maps color name strings to colors (discrete to discrete).
- `scale_color_continuous` which linearly maps from some bounded continuous space to colors (continuous to continuous).

At least one user in at least one case was trying to use `scale_color_continuous` to convert a _categorical_ integer field into a discrete set of colors. This makes me think we're missing one more scale:

- `scale_color_categorical` which maps some subset of a (possibly continuous) space to evenly spaced colors. This scale should absolutely participate in legends and facets because we're ignoring the continuity of the type and treating these values as discrete/categorical.

Getting back to your question: I thought scale_color_continuous was scale_color_categorical, ergo all these changes and why I thought the output of scale_color_continuous is "discrete" (aka categorical).

I guess this means I should create a new scale?

Reply:
Ah, that's helpful context.

So, here's a concrete example of bad behavior:
```python3
import hail as hl
ht = hl.utils.range_table(10)
ht = ht.annotate(subpop = ht.idx % 3)
ht = ht.annotate(proj_title = hl.literal(['a', 'b'])[ht.idx % 2])
ht = ht.annotate(x = hl.rand_int32(0, 3))

p = hl.ggplot.ggplot(ht, hl.ggplot.aes(x=ht.x, color =  ht.subpop))
p += hl.ggplot.geom_bar()
p += hl.ggplot.xlab('x')
p += hl.ggplot.ylab('count')
p += hl.ggplot.ggtitle(" Subcontinental PCA pre-outlier removal")
p += hl.ggplot.labs(shape = 'Project', color = 'Population')

p
```
This should create a stacked bar plot with color distinguishing between the different populations. Instead, it gives me this black bar plot.

<img width="771" alt="Screen Shot 2023-04-07 at 12 39 09" src="https://user-images.githubusercontent.com/106194/230645160-3a2435bf-6130-4d98-a25d-d6c73c14d12a.png">

The trouble is that `geom_bar` with a continuous color scaling doesn't make sense.

Reply:
The tricker case is `geom_point`. `geom_point` *does* have a reasonable interpretation of continuous color, but:
1. We don't produce a continuous visual legend/scale for that color.
2. In reality, we want to treat this integer as a categorical and get a distinct legend entry for each value.
<img width="741" alt="Screen Shot 2023-04-07 at 12 42 01" src="https://user-images.githubusercontent.com/106194/230645538-2d82bf1f-72ab-4200-a2c8-bb378fbe2060.png">

Reply:
I can't specify `scales` in the `ggplot` function. If I try to manually specify `scale={'color': hl.ggplot.scale_color_discrete()}` by directly calling the class constructor, Hail's ggplot complains:

```
ValueError: Invalid scale for aesthetic color of type int32
```

Reply:
Ah, thanks, I understand now. I think the core issue is that we don't have any way to override the discrete/continuous inference from datatype. This works, but I agree we should find a way to do it more directly:
```
import hail as hl
ht = hl.utils.range_table(10)
ht = ht.annotate(subpop = hl.str(ht.idx % 3))
ht = ht.annotate(proj_title = hl.literal(['a', 'b'])[ht.idx % 2])
ht = ht.annotate(x = hl.rand_int32(0, 3))

p = hl.ggplot.ggplot(ht, hl.ggplot.aes(x=ht.x, fill =  ht.subpop))
p += hl.ggplot.geom_bar()
p += hl.ggplot.xlab('x')
p += hl.ggplot.ylab('count')
p += hl.ggplot.ggtitle(" Subcontinental PCA pre-outlier removal")
p += hl.ggplot.labs(shape = 'Project', color = 'Population')

p.show()
```

We should copy ggplot's interface for manually specifying continuous/discrete, but after a few minutes I haven't been able to figure out how to do it. I'll let you know if I find it.

Reply:
It's possible ggplot lets you assign a discrete scale to any datatype, in which case we should too.

Reply:
Ah, R has a generic way to force a continuous type to be used categorically, `factor`. So they would write this
```
aes(x=ht.x, fill = factor(ht.subpop))
```
Seems like a good pattern to emulate.